torch_geometric/nn/conv/mf_conv.py

```diff
@@ -96,15 +91,45 @@ def forward(self, x: Union[Tensor, OptPairTensor], edge_index: Adj,
         h = self.propagate(edge_index, x=x, size=size)
 
         out = h.new_empty(list(h.size())[:-1] + [self.out_channels])
-        for i, (lin_l, lin_r) in enumerate(zip(self.lins_l, self.lins_r)):
-            idx = (deg == i).nonzero().view(-1)
-            r = lin_l(h.index_select(self.node_dim, idx))
-
-            if x_r is not None:
-                r = r + lin_r(x_r.index_select(self.node_dim, idx))
-
-            out.index_copy_(self.node_dim, idx, r)
 
+        # idx select loop for l
+        h_sel_list, type_list_l, idx_list = [], [], []
+        for i in range(self.max_degree + 1):
+            idx_i = (deg == i).nonzero().view(-1)
+            h_idx_sel = h.index_select(self.node_dim, idx_i)
+            idx_list.append(idx_i)
+            N = h_idx_sel.size(0)
+            h_sel_list.append(h_idx_sel)
+            type_list_l.append(torch.full((N, ), i, dtype=torch.long))
+        x_l = torch.cat(h_sel_list, dim=0)
+        type_vec_l = torch.cat(type_list_l, dim=0)
+        idx = torch.cat(idx_list, dim=0)
+
+        # apply lin_l
+        r = self.lin_l(x_l, type_vec_l)
+
+        # idx select loop for r
+        if x_r is not None:
+            r_sel_list, type_list_r = [], []
+            count = 0
+            for i in range(self.max_degree + 1):
+                idx_i = idx_list[i]
+
+                N = idx_i.numel()
+                if N == 0:
+                    continue
+                r_idx_sel = x_r.index_select(self.node_dim, idx_i)
+                r_sel_list.append(r_idx_sel)
+                idx_list.append(idx_i + count)
+                count += N
+                type_list_r.append(torch.full((N, ), i, dtype=torch.long))
+            x_r = torch.cat(r_sel_list, dim=0)
+            type_vec_r = torch.cat(type_list_r, dim=0)
```

Comment:
Isn't `type_vec_r` the same as `type_vec_l`?

Reply:
oh yeah i updated the loops to use the same idx list but forgot to remove the identical type vec/lists